Major updates

reinhold-willcox opened the pull request:
This PR incorporate many distinct changes (apologies to those who prefer incremental PRs), but I think the final product is close to what we've been working towards.
     
New features/changes:    
     
- created run_sw.py file in the source code, which contains the parts of the interface file which do not require user input (the SW calls and configure_code_run)    
- added RANDOM_SEED parameter which increments correctly    
- added --random-seed as first parameter in the gridfiles    
- added a stroopwafel_interface file (which will later be removed and put into COMPAS) with associated stoopwafel_dev softlink to make testing easier.    
- cleaned up formatting and directory structure for the interface file (testing now requires a SW_ROOT env var)    
- removed support for pythonSubmit file - now all parameters are set in SW, or take on the default COMPAS values    
- added time request parameter for user to specify the requested walltime on hpc systems 
- added line to explicitly create output folder if it does not exist (resolves known issue for some operating systems)    
- added call to postprocessing script so that COMPAS_Output.h5 is created automatically with the SW call    
     
Remaining features to be addressed:    
     
- how to set the default value for num_per_batch - should this be the same for mc and non-mc runs? 

Edit: this supersedes PR #6, so I'm closing that one

reinhold-willcox committed:
removed excess stroopwafel_interface file

reinhold-willcox committed:
added current best version of stroopwafel interface, and the run_sw file

reinhold-willcox committed:
added --random-seed to output

Commit:
added default stroopwafel_interface file and stroopwafel_injections (for nskicks projects) to principle_branch

reinhold-willcox committed:
fixed range for remnant masses from sw injections

Commit:
added schneider interface w/ pysub

reinhold-willcox committed:
removed many commented lines

Commit:
fixed implementation of random seed for arbitrary base seed

Commit:
fixed bug where pythonSubmit args were not being passed to the slurm files, and added check to mkdir if notexists(dir) in process_batches

reinhold-willcox committed:
added more parameters to sw_injections and related files to remove the need for a pythonSubmit file

reinhold-willcox committed:
added wall-time request for hpc

Commit:
added default time_request

reinhold-willcox committed:
fixed random seed for np random number generator

reinhold-willcox committed:
updating stroopwafel interface files for my projects

reinhold-willcox committed:
reset upper mass limit to 150 Msun

reinhold-willcox committed:
fixed calls to import constants to be more explicit

reinhold-willcox committed:
removed pythonSubmit import and functionality, everything can now be done in SW

reinhold-willcox committed:
renamed param 'num systems per core' to 'num systems per batch'

reinhold-willcox committed:
removed 'if not mc_only' requirement for AIS methods - this is handled internally

reinhold-willcox committed:
cleaned up formatting of sw interface, and added call to postprocessing file

reinhold-willcox committed:
moved interface file to 'tests' dir and hid Reinhold_tests dir

reinhold-willcox committed:
improved comments and function definitions in sw interface

reinhold-willcox committed:
added time_request parameter to all run_code calls, including genais, and added time_request param to genais

reinhold-willcox committed:
renamed dns to dco (more accurate), and added comment on how to find COMPAS defaults

reinhold-willcox committed:
added sw int

reinhold-willcox committed:
updates to injections and schneider files

reinhold-willcox committed:
added flag for mc_only back in

reinhold-willcox commented:
@floriskummer I had to put the mc_only restriction back in because I got an error without it. I'm not sure if the error has to do with what the backend ignores when mc_only=True, or if it occurs for mc_only=False as well, but its related to the transform_variables function.

```
Traceback (most recent call last):
  File "stroopwafel_interface.py", line 231, in <module>
    selection_effects, rejected_systems)
  File "/home/rwillcox/astro/stroopwafel/stroopwafel_dev/run_sw.py", line 72, in run_stroopwafel
    sw_object.postprocess(distributions.Gaussian, only_hits = False) #Run it to create weights, if you want only hits in the output, then make only_hits = True
  File "/home/rwillcox/astro/stroopwafel/stroopwafel_dev/sw.py", line 234, in postprocess
    [location.transform_variables_to_new_scales() for location in locations]
  File "/home/rwillcox/astro/stroopwafel/stroopwafel_dev/sw.py", line 234, in <listcomp>
    [location.transform_variables_to_new_scales() for location in locations]
  File "/home/rwillcox/astro/stroopwafel/stroopwafel_dev/classes.py", line 114, in transform_variables_to_new_scales
    self.dimensions[dimension] = np.log10(value)
```

floriskummer commented:
Hmm.. I have not seen this error before. I tried running it as well and it worked for both mc and sw (but this is from the new_updates branch). Do you also have the type of error? It looks like something goes wrong with storing the dimensions (although I wouldn't have the faintest idea why)

stellarphysicsandpopulations commented:
Hmm.  Yes, in general I'd prefer more incremental changes.  But at first sight it looks like your intents in these changes are good.  I can certainly understand why you'd be impatient to get moving.    Thanks again for all the thought and work you've been putting in. 

Since you closed PR6, allow me to restate one of the themes from Floris' message:  
    Please be extra-careful when changing anything inside the "stroopwafel" directory, i.e., not the interface but the core internal files Loki wrote.  (And, as far as possible, please try to stick to Loki's coding style -- e.g., in the way "constants" are called -- unless you are absolutely certain that changing is better.)
    Multiple times now I've wrongly thought that I caught a bug in those methods, only to spend more time thinking/reading and to later realise that Loki was doing things in a more elegant way than I first thought.
    
(That doesn't mean we can't change anything. For example, as above, I think some of the naming can probably be clarified. But in those files please be extra careful, even in your own branches.)

reinhold-willcox commented:
@floriskummer Yes, I should have included the TypeError above. Here it is: 
```
TypeError: ufunc 'log10' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

@stellarphysicsandpopulations You're absolutely right, I have also discovered several times that Loki's original way of writing something is cleverer than my quick bugfix. In this case, I had modified the constants call in order to make the random-seed functionality work, but this broke the call to the other constants. I added back in the original call (as well as the new, random-seed one) and things seem to be working again.

reinhold-willcox committed:
updated final command to merge the h5's back together. Latest changes in compas now output directly to h5, so this change incorporates the new h5copy recombination file.

reinhold-willcox commented:
The latest change uses the new h5copy file in COMPAS, which merges the individual h5 outputs (which are now directly outputted from COMPAS) into one main one. 

Separately, what are everyone's thoughts on the current state of this PR? Is it ready to be merged in, or are there outstanding issues? I understand the SW functionality is still not quite ready, but I'm not sure I'm equipped to help sort that out.

reinhold-willcox committed:
removed obsolete command line arg 'logfile delimter COMMA'

floriskummer commented:
Great Reinhold! I have been a bit busy with interview preparations this week, but I will try to have a look at the final changes by the end of this week.

Commit:
reads in the new h5 files in the interface

reinhold-willcox committed:
added h5py import to sw interface, and added in default params from pythonSubmit.py, which are the compas fiducial parameters

Commit:
removed references to only locally relevant directory

reinhold-willcox committed:
Deleted irrelevant local file

reinhold-willcox committed:
Deleted irrelevant local file
